src/bun.js/node/node_fs_watcher.zig

```diff
@@ -540,241 +562,202 @@ pub const FSWatcher = struct {
+                    else if (this.encoding == .utf8) {
+                        filename = JSC.ZigString.fromUTF8(file_name).toValueGC(this.globalThis);
+                    } else {
+                        // convert to desired encoding
+                        filename = Encoder.toStringAtRuntime(file_name.ptr, file_name.len, this.globalThis, this.encoding);
                     }
                 }
+                var args = [_]JSC.JSValue{
+                    JSC.ZigString.static(eventType).toValue(this.globalThis),
+                    filename,
+                };
+                _ = listener.callWithGlobalThis(
+                    this.globalThis,
+                    &args,
+                );
             }
         }
+    }
 
-        pub fn ref(this: *JSObject) void {
-            if (this.closed) return;
-
-            if (!this.persistent) {
-                this.persistent = true;
-                this.poll_ref.ref(this.getFSWatcher().ctx);
-            }
+    pub fn doRef(this: *FSWatcher, _: *JSC.JSGlobalObject, _: *JSC.CallFrame) callconv(.C) JSC.JSValue {
+        if (!this.isClosed() and !this.persistent) {
+            this.persistent = true;
+            this.poll_ref.ref(this.ctx);
         }
+        return JSC.JSValue.jsUndefined();
+    }
 
-        pub fn doRef(this: *JSObject, _: *JSC.JSGlobalObject, _: *JSC.CallFrame) callconv(.C) JSC.JSValue {
-            this.ref();
-            return JSC.JSValue.jsUndefined();
+    pub fn doUnref(this: *FSWatcher, _: *JSC.JSGlobalObject, _: *JSC.CallFrame) callconv(.C) JSC.JSValue {
+        if (this.persistent) {
+            this.persistent = false;
+            this.poll_ref.unref(this.ctx);
         }
+        return JSC.JSValue.jsUndefined();
+    }
 
-        pub fn unref(this: *JSObject) void {
-            if (this.persistent) {
-                this.persistent = false;
-                this.poll_ref.unref(this.getFSWatcher().ctx);
-            }
-        }
+    pub fn hasRef(this: *FSWatcher, _: *JSC.JSGlobalObject, _: *JSC.CallFrame) callconv(.C) JSC.JSValue {
+        return JSC.JSValue.jsBoolean(this.persistent);
+    }
 
-        pub fn doUnref(this: *JSObject, _: *JSC.JSGlobalObject, _: *JSC.CallFrame) callconv(.C) JSC.JSValue {
-            this.unref();
-            return JSC.JSValue.jsUndefined();
-        }
+    // this can be called from Watcher Thread or JS Context Thread
+    pub fn refTask(this: *FSWatcher) bool {
+        // stop new references
+        if (this.isClosed()) return false;
+        _ = this.task_count.fetchAdd(1, .Monotonic);
+        return true;
+    }
 
-        pub fn hasRef(this: *JSObject, _: *JSC.JSGlobalObject, _: *JSC.CallFrame) callconv(.C) JSC.JSValue {
-            return JSC.JSValue.jsBoolean(this.persistent);
-        }
+    pub fn hasPendingActivity(this: *FSWatcher) callconv(.C) bool {
+        @fence(.Acquire);
+        return this.task_count.load(.Acquire) > 0 or !this.closed.load(.Acquire);
```

Comment:
This has a race condition. task_count could become greater than zero after this.closed is loaded, or vice versa

Reply:
if user call `.close()` but we still have pending tasks I just wait and deinit after it, if is not closed and have no tasks we need to keep it alive, so will only close when user call `.close()` and we have no tasks

Reply:
The `finalize` function will be called (on the main thread) when `hasPendingActivity` (called on a different thread) returns false. It will be called only once per instance. Instead of having two atomics, why not have one for pending activity specifically?

Reply:
The race condition is when task_count becomes > 0 at the same time as the developer calls `.close()` from JS

Reply:
It is possible for that to happen because task_count is incremented on a different thread

Reply:
> The `finalize` function will be called (on the main thread) when `hasPendingActivity` (called on a different thread) returns false. It will be called only once per instance. Instead of having two atomics, why not have one for pending activity specifically?

We can have multiple tasks pending, I can just start task_count with 1 and close will just decrement it) so will be only one atomic, and I rename it back to ref_count. I update this.

Reply:
> The `finalize` function will be called (on the main thread) when `hasPendingActivity` (called on a different thread) returns false. It will be called only once per instance. Instead of having two atomics, why not have one for pending activity specifically?

We can have multiple pending tasks, if user calls close and I just decrement the ref_count more pending tasks can be created (because ref_count > 0), if I use `closed` flag I can stop the creation of new tasks and end the process sooner.

Reply:
fixed into https://github.com/oven-sh/bun/pull/3396/commits/feb30ca7197f17b39c0057dc6a0c455b2964ff89